test/test_image.py

```diff
@@ -54,176 +53,205 @@ def normalize_dimensions(img_pil):
-        with get_tmp_dir() as d:
-            fname, content = 'test1.bin', b'TorchVision\211\n'
-            fpath = os.path.join(d, fname)
-            content_tensor = torch.tensor(list(content), dtype=torch.uint8)
-            write_file(fpath, content_tensor)
-
-            with open(fpath, 'rb') as f:
-                saved_content = f.read()
-            self.assertEqual(content, saved_content)
-            os.unlink(fpath)
-
-    def test_write_file_non_ascii(self):
-        with get_tmp_dir() as d:
-            fname, content = '日本語(Japanese).bin', b'TorchVision\211\n'
-            fpath = os.path.join(d, fname)
-            content_tensor = torch.tensor(list(content), dtype=torch.uint8)
-            write_file(fpath, content_tensor)
-
-            with open(fpath, 'rb') as f:
-                saved_content = f.read()
-            self.assertEqual(content, saved_content)
-            os.unlink(fpath)
+@pytest.mark.parametrize('img_path', [
+    pytest.param(jpeg_path, id=_get_safe_image_name(jpeg_path))
+    for jpeg_path in get_images(IMAGE_ROOT, ".jpg")
+])
+@pytest.mark.parametrize('pil_mode, mode', [
+    (None, ImageReadMode.UNCHANGED),
+    ("L", ImageReadMode.GRAY),
+    ("RGB", ImageReadMode.RGB),
+])
+def test_decode_jpeg(img_path, pil_mode, mode):
+
+    with Image.open(img_path) as img:
+        is_cmyk = img.mode == "CMYK"
+        if pil_mode is not None:
+            if is_cmyk:
+                # libjpeg does not support the conversion
+                pytest.xfail("Decoding a CMYK jpeg isn't supported")
+            img = img.convert(pil_mode)
+        img_pil = torch.from_numpy(np.array(img))
+        if is_cmyk:
+            # flip the colors to match libjpeg
+            img_pil = 255 - img_pil
+
+    img_pil = normalize_dimensions(img_pil)
+    data = read_file(img_path)
+    img_ljpeg = decode_image(data, mode=mode)
+
+    # Permit a small variation on pixel values to account for implementation
+    # differences between Pillow and LibJPEG.
+    abs_mean_diff = (img_ljpeg.type(torch.float32) - img_pil).abs().mean().item()
+    assert abs_mean_diff < 2
+
+
+def test_decode_jpeg_errors():
+    with pytest.raises(RuntimeError, match="Expected a non empty 1-dimensional tensor"):
+        decode_jpeg(torch.empty((100, 1), dtype=torch.uint8))
+
+    with pytest.raises(RuntimeError, match="Expected a torch.uint8 tensor"):
+        decode_jpeg(torch.empty((100,), dtype=torch.float16))
+
+    with pytest.raises(RuntimeError, match="Not a JPEG file"):
+        decode_jpeg(torch.empty((100), dtype=torch.uint8))
+
+
+def test_decode_bad_huffman_images():
+    # sanity check: make sure we can decode the bad Huffman encoding
+    bad_huff = read_file(os.path.join(DAMAGED_JPEG, 'bad_huffman.jpg'))
+    decode_jpeg(bad_huff)
+
+
+@pytest.mark.parametrize('img_path', [
+    pytest.param(truncated_image, id=_get_safe_image_name(truncated_image))
+    for truncated_image in glob.glob(os.path.join(DAMAGED_JPEG, 'corrupt*.jpg'))
```

Comment:
```test/test_image.py::test_damaged_corrupt_images[corrupt.jpg] FAILED      [ 77%]```

@NicolasHug , This image raise a different error comparing the other images, I will fix this soon. And how do you think this problem should be solved?

Reply:
Perhaps see if you can build a regexp that accept either one of those messages: https://docs.pytest.org/en/6.2.x/reference.html#pytest.raises

You can also define 2 messages and use one or the other depending on img_path?

No need to spend too much time on this though, it's not worth it: we can also just remove the `match`.

Reply:
Ok, I just wrote a conditional expression to select the error messages.

```python
    if 'corrupt34' in img_path:
        match_message = "Image is incomplete or truncated"
    else:
        match_message = "Unsupported marker type"
```